README.md
# Changes

On Android, the presense check in enableReaderMode will cause problems in DESFire mutual authentication. The timeout is set to 10 seconds to prevent interruption of authentication

# Flutter NFC Kit

[![pub version](https://img.shields.io/pub/v/flutter_nfc_kit)](https://pub.dev/packages/flutter_nfc_kit)
![Build Example App](https://github.com/nfcim/flutter_nfc_kit/workflows/Build%20Example%20App/badge.svg)

Yet another plugin to provide NFC functionality on Android and iOS.

This plugin's functionalities include:

* read metadata and read & write NDEF records of tags / cards complying with:
  * ISO 14443 Type A & Type B (NFC-A / NFC-B / MIFARE Classic / MIFARE Plus / MIFARE Ultralight / MIFARE DESFire)
  * ISO 18092 (NFC-F / FeliCa)
  * ISO 15963 (NFC-V)
* transceive commands with tags / cards complying with:
  * ISO 7816 Smart Cards (layer 4, in APDUs)
  * other device-supported technologies (layer 3, in raw commands, Android only)

Note that due to API limitations not all operations are supported on both platforms.

This library uses [ndef](https://pub.dev/packages/ndef) for NDEF record encoding & decoding.

## Setup

Thank [nfc_manager](https://pub.dev/packages/nfc_manager) plugin for these instructions.

### Android

* Add [android.permission.NFC](https://developer.android.com/reference/android/Manifest.permission.html#NFC) to your `AndroidManifest.xml`.

### iOS

* Add [Near Field Communication Tag Reader Session Formats Entitlements](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_nfc_readersession_formats) to your entitlements.

* Add [NFCReaderUsageDescription](https://developer.apple.com/documentation/bundleresources/information_property_list/nfcreaderusagedescription) to your `Info.plist`.

* Add [com.apple.developer.nfc.readersession.felica.systemcodes](https://developer.apple.com/documentation/bundleresources/information_property_list/systemcodes) and [com.apple.developer.nfc.readersession.iso7816.select-identifiers](https://developer.apple.com/documentation/bundleresources/information_property_list/select-identifiers) to your `Info.plist` as needed. WARNING: you **MUST** add them before invoking `poll` with `readIso18092` or `readIso15693` enabled, or your NFC **WILL BE TOTALLY UNAVAILABLE BEFORE REBOOT** due to a [CoreNFC bug](https://github.com/nfcim/flutter_nfc_kit/issues/23).

* Open Runner.xcworkspace with Xcode and navigate to project settings then the tab _Signing & Capabilities._

* Select the Runner in targets in left sidebar then press the "+ Capability" in the left upper corner and choose _Near Field Communication Tag Reading._

## Usage

Simple example:

```dart
import 'package:flutter_nfc_kit/flutter_nfc_kit.dart';
import 'package:ndef/ndef.dart' as ndef;

var availability = await FlutterNfcKit.nfcAvailability;
if (availability != NFCAvailability.available) {
    // oh-no
}

// timeout only works on Android, while the latter two messages are only for iOS
var tag = await FlutterNfcKit.poll(timeout: Duration(seconds: 10),
  iosMultipleTagMessage: "Multiple tags found!", iosAlertMessage: "Scan your tag");

print(jsonEncode(tag));
if (tag.type == NFCTagType.iso7816) {
    var result = await FlutterNfcKit.transceive("00B0950000", Duration(seconds: 5)); // timeout is still Android-only, persist until next change
    print(result);
}
// iOS only: set alert message on-the-fly
// this will persist until finish()
await FlutterNfcKit.setIosAlertMessage("hi there!");

// read NDEF records if available
if (tag.ndefAvailable){
  /// decoded NDEF records (see [ndef.NDEFRecord] for details)
  /// `UriRecord: id=(empty) typeNameFormat=TypeNameFormat.nfcWellKnown type=U uri=https://github.com/nfcim/ndef`
  for (var record in await FlutterNfcKit.readNDEFRecords(cached: false)) {
    print(record.toString());
  }
  /// raw NDEF records (data in hex string)
  /// `{identifier: "", payload: "00010203", type: "0001", typeNameFormat: "nfcWellKnown"}`
  for (var record in await FlutterNfcKit.readNDEFRawRecords(cached: false)) {
    print(jsonEncode(record).toString());
  }
}

// write NDEF records if applicable
if (tag.ndefWritable) {
  // decoded NDEF records
  await FlutterNfcKit.writeNDEFRecords([new ndef.UriRecord.fromUriString("https://github.com/nfcim/flutter_nfc_kit")]);
  // raw NDEF records
  await FlutterNfcKit.writeNDEFRawRecords([new NDEFRawRecord("00", "0001", "0002", "0003", ndef.TypeNameFormat.unknown)]);
}

// Call finish() only once
await FlutterNfcKit.finish();
// iOS only: show alert/error message on finish
await FlutterNfcKit.finish(iosAlertMessage: "Success");
// or
await FlutterNfcKit.finish(iosErrorMessage: "Failed");
```

A more complicated example can be seen in `example` dir.

Refer to the [documentation](https://pub.dev/documentation/flutter_nfc_kit/) for more information.

### Error codes

We use error codes with similar meaning as HTTP status code. Brief explanation and error cause in string (if available) will also be returned when an error occurs.


CHANGELOG.md
## 0.0.1

* Initial release
* Support reading metadata of cards of standard ISO 14443-4 Type A & Type B (NFC-A / NFC-B / Mifare Classic / Mifare Ultralight)
* Support transceiving APDU with cards of standard ISO 7816

## 0.0.2

* Support reading metadata of cards of standard ISO 18092 / JIS 6319 (NFC-F / Felica)
* Support reading metadata of cards of standard ISO 15963 (NFC-V)
* Support reading GUID of China ID Card (which is non-standard)
* Add more documentation
* Release with MIT License

## 0.0.3

* Fix compilation errors of iOS plugin

## 0.0.4

* Fix IsoDep unintentionally closing of the Android plugin
* Fix incorrect standards description

## 0.0.5

* Fix finish method blocking error

## 0.0.6

* Avoid returning redundant result if user finish the read operation prematurely

## 0.0.7

* Allow data transceive on lower layer (Android only)

## 1.0.0

* Remove China ID Card support due to support of lower layer transceiving
* Fix some racing problem on iOS
* We are out-of beta and releasing on time!

## 1.0.1

* Fix IllegalStateException & Add MethodResultWrapper (Thanks to @smlu)

## 1.0.2

* Remove redundant code in Android plugin
* Format dart code

## 1.1.0

* Add NFC session alert/error message on iOS (Thanks to @smlu)
* Support execution timeout as optional parameter on Android for `poll` and `transceive` (Thanks to @smlu)
* Accept command in the type of hex string / `UInt8List` in `transceive` and return in the same type (Thanks to @smlu)

## 1.2.0

* Add support for NFC 18902 on iOS
* Add initial NDEF support (read only, no decoding)
* Allow disabling platform sound on Android when polling card

## 2.0.0

* Switch to [ndef](https://pub.dev/packages/ndef) for NDEF record encoding & decoding (breaking API change)
* Support writing NDEF records
* Add NDEF writing in example app

## 2.0.1

* Fix compiling problem on iOS

## 2.0.2

* Fix format of CHANGELOG.md
* Format source code to pass static analysis

## 2.1.0

* Update to latest version of `ndef` package
* Update example app on writing NDEF records

## 2.2.0

* Allow specifying needed NFC technologies in `poll` (fix [#15](https://github.com/nfcim/flutter_nfc_kit/issues/15))

## 2.2.1

* Disable ISO 18092 in `poll` by default due to iOS CoreNFC bug (see [#23](https://github.com/nfcim/flutter_nfc_kit/issues/23))
* Bump dependencies & fix some deprecation warnings

## Custom version

* Set enableReaderMode presenseCheck to 10 seconds timeout to prevent interruption of authentication

android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt
package im.nfc.flutter_nfc_kit

import android.app.Activity
import android.nfc.FormatException
import android.nfc.NdefMessage
import android.nfc.NdefRecord
import android.nfc.NfcAdapter
import android.nfc.NfcAdapter.*
import android.nfc.tech.*
import android.os.Handler
import android.os.Looper
import im.nfc.flutter_nfc_kit.ByteUtils.hexToBytes
import im.nfc.flutter_nfc_kit.ByteUtils.toHexString
import io.flutter.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.util.*
import kotlin.concurrent.schedule
import android.os.Bundle

class FlutterNfcKitPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    companion object {
        private val TAG = FlutterNfcKitPlugin::class.java.name
        private var activity: Activity? = null
        private var pollingTimeoutTask: TimerTask? = null
        private var tagTechnology: TagTechnology? = null
        private var ndefTechnology: Ndef? = null

        private fun TagTechnology.transceive(data: ByteArray, timeout: Int?): ByteArray {
            if (timeout != null) {
                try {
                    val timeoutMethod = this.javaClass.getMethod("setTimeout", Int::class.java)
                    timeoutMethod.invoke(this, timeout)
                } catch (ex: Throwable) {
                }
            }
            val transceiveMethod = this.javaClass.getMethod("transceive", ByteArray::class.java)
            return transceiveMethod.invoke(this, data) as ByteArray
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_nfc_kit")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        handleMethodCall(call, MethodResultWrapper(result))
    }

    private fun handleMethodCall(call: MethodCall, result: Result) {

        val nfcAdapter = getDefaultAdapter(activity)

        if (nfcAdapter?.isEnabled != true && call.method != "getNFCAvailability") {
            result.error("404", "NFC not available", null)
            return
        }

        when (call.method) {

            "getNFCAvailability" -> {
                when {
                    nfcAdapter == null -> result.success("not_supported")
                    nfcAdapter.isEnabled -> result.success("available")
                    else -> result.success("disabled")
                }
            }

            "poll" -> {
                val timeout = call.argument<Int>("timeout")!!
                // technology and option bits are set in Dart code
                val technologies = call.argument<Int>("technologies")!!
                pollTag(nfcAdapter, result, timeout, technologies)
            }

            "finish" -> {
                pollingTimeoutTask?.cancel()
                try {
                    val tagTech = tagTechnology
                    if (tagTech != null && tagTech.isConnected) {
                        tagTech.close()
                    }
                    val ndefTech = ndefTechnology
                    if (ndefTech != null && ndefTech.isConnected) {
                        ndefTech.close()
                    }
                } catch (ex: IOException) {
                    Log.e(TAG, "Close tag error", ex)
                }
                nfcAdapter.disableReaderMode(activity)
                result.success("")
            }

            "transceive" -> {
                val tagTech = tagTechnology
                val req = call.argument<Any>("data")
                if (req == null || (req !is String && req !is ByteArray)) {
                    result.error("400", "Bad argument", null)
                    return
                }
                if (tagTech == null) {
                    result.error("406", "No tag polled", null)
                    return
                }

                try {
                    if (!tagTech.isConnected) {
                        val ndefTech = ndefTechnology
                        // close previously connected technology
                        if (ndefTech !== null && ndefTech.isConnected) {
                            ndefTech.close()
                        }
                        tagTech.connect()
                    }
                    val sendingBytes = when (req) {
                        is String -> req.hexToBytes()
                        else -> req as ByteArray
                    }
                    val timeout = call.argument<Int>("timeout")
                    val resp = tagTech.transceive(sendingBytes, timeout)
                    when (req) {
                        is String -> result.success(resp.toHexString())
                        else -> result.success(resp)
                    }
                } catch (ex: IOException) {
                    Log.e(TAG, "Transceive Error: $req", ex)
                    result.error("500", "Communication error", ex.localizedMessage)
                } catch (ex: InvocationTargetException) {
                    Log.e(TAG, "Transceive Error: $req", ex.cause ?: ex)
                    result.error("500", "Communication error", ex.cause?.localizedMessage)
                } catch (ex: IllegalArgumentException) {
                    Log.e(TAG, "Command Error: $req", ex)
                    result.error("400", "Command format error", ex.localizedMessage)
                } catch (ex: NoSuchMethodException) {
                    Log.e(TAG, "Transceive not supported: $req", ex)
                    result.error("405", "Transceive not supported for this type of card", null)
                }
            }

            "readNDEF" -> {
                if (ndefTechnology == null) {
                    if (tagTechnology == null) {
                        result.error("406", "No tag polled", null)
                    } else {
                        result.error("405", "NDEF not supported on current tag", null)
                    }
                    return
                }
                val ndef = ndefTechnology!!
                try {
                    // try to connect
                    if (!ndef.isConnected) {
                        val otherTech = tagTechnology
                        // close previously connected technology
                        if (otherTech !== null && otherTech.isConnected) {
                            otherTech.close()
                        }
                        ndef.connect()
                    }
                    // read NDEF message
                    val message: NdefMessage? = if (call.argument<Boolean>("cached")!!) {
                        ndef.cachedNdefMessage
                    } else {
                        ndef.ndefMessage
                    }
                    val parsedMessages = mutableListOf<Map<String, String>>()
                    if (message != null) {
                        for (record in message.records) {
                            parsedMessages.add(mapOf(
                                    "identifier" to record.id.toHexString(),
                                    "payload" to record.payload.toHexString(),
                                    "type" to record.type.toHexString(),
                                    "typeNameFormat" to when (record.tnf) {
                                        NdefRecord.TNF_ABSOLUTE_URI -> "absoluteURI"
                                        NdefRecord.TNF_EMPTY -> "empty"
                                        NdefRecord.TNF_EXTERNAL_TYPE -> "nfcExternal"
                                        NdefRecord.TNF_WELL_KNOWN -> "nfcWellKnown"
                                        NdefRecord.TNF_MIME_MEDIA -> "media"
                                        NdefRecord.TNF_UNCHANGED -> "unchanged"
                                        else -> "unknown"
                                    }
                            ))
                        }
                    }
                    result.success(JSONArray(parsedMessages).toString())
                } catch (ex: IOException) {
                    Log.e(TAG, "Read NDEF Error", ex)
                    result.error("500", "Communication error", ex.localizedMessage)
                } catch (ex: FormatException) {
                    Log.e(TAG, "NDEF Format Error", ex)
                    result.error("400", "NDEF format error", ex.localizedMessage)
                }
            }

            "writeNDEF" -> {
                if (ndefTechnology == null) {
                    if (tagTechnology == null) {
                        result.error("406", "No tag polled", null)
                    } else {
                        result.error("405", "NDEF not supported on current tag", null)
                    }
                    return
                }
                val ndef = ndefTechnology!!
                if(ndef.isWritable() == false){
                    result.error("405", "Tag not writable", null)
                }
                try {
                    // try to connect
                    if (!ndef.isConnected) {
                        val otherTech = tagTechnology
                        // close previously connected technology
                        if (otherTech !== null && otherTech.isConnected) {
                            otherTech.close()
                        }
                        ndef.connect()
                    }
                    // generate NDEF message
                    val jsonString = call.argument<String>("data")!!
                    val recordData = JSONArray(jsonString)
                    val records = Array<NdefRecord>(recordData.length(),init= {i:Int -> 
                        val record:JSONObject = recordData.get(i) as JSONObject
                        NdefRecord(
                            when(record.getString("typeNameFormat")){
                                "absoluteURI" -> NdefRecord.TNF_ABSOLUTE_URI
                                "empty" -> NdefRecord.TNF_EMPTY
                                "nfcExternal" -> NdefRecord.TNF_EXTERNAL_TYPE
                                "nfcWellKnown" -> NdefRecord.TNF_WELL_KNOWN
                                "media" -> NdefRecord.TNF_MIME_MEDIA
                                "unchanged" -> NdefRecord.TNF_UNCHANGED
                                else -> NdefRecord.TNF_UNKNOWN
                            },
                            record.getString("type").hexToBytes(),
                            record.getString("identifier").hexToBytes(),
                            record.getString("payload").hexToBytes()
                        )})
                    // write NDEF message
                    val message: NdefMessage = NdefMessage(records)
                    ndef.writeNdefMessage(message)
                    result.success("")
                } catch (ex: IOException) {
                    Log.e(TAG, "Write NDEF Error", ex)
                    result.error("500", "Communication error", ex.localizedMessage)
                } catch (ex: FormatException) {
                    Log.e(TAG, "NDEF Format Error", ex)
                    result.error("400", "NDEF format error", ex.localizedMessage)
                }
            }

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {}

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        if (activity != null) return
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        pollingTimeoutTask?.cancel()
        pollingTimeoutTask = null
        tagTechnology = null
        ndefTechnology = null
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}

    override fun onDetachedFromActivityForConfigChanges() {}

    private fun pollTag(nfcAdapter: NfcAdapter, result: Result, timeout: Int, technologies: Int) {

        pollingTimeoutTask = Timer().schedule(timeout.toLong()) {
            nfcAdapter.disableReaderMode(activity)
            result.error("408", "Polling tag timeout", null)
        }

        var options : Bundle = Bundle()
        options.putInt(NfcAdapter.EXTRA_READER_PRESENCE_CHECK_DELAY, 10000)

        nfcAdapter!!.enableReaderMode(activity, { tag ->
            pollingTimeoutTask?.cancel()

            // common fields
            val type: String
            val id = tag.id.toHexString()
            val standard: String
            // ISO 14443 Type A
            var atqa = ""
            var sak = ""
            // ISO 14443 Type B
            var protocolInfo = ""
            var applicationData = ""
            // ISO 7816
            var historicalBytes = ""
            var hiLayerResponse = ""
            // NFC-F / Felica
            var manufacturer = ""
            var systemCode = ""
            // NFC-V
            var dsfId = ""
            // NDEF
            var ndefAvailable = false
            var ndefWritable = false
            var ndefCanMakeReadOnly = false
            var ndefCapacity = 0
            var ndefType = ""

            if (tag.techList.contains(NfcA::class.java.name)) {
                val aTag = NfcA.get(tag)
                atqa = aTag.atqa.toHexString()
                sak = byteArrayOf(aTag.sak.toByte()).toHexString()
                tagTechnology = aTag
                when {
                    tag.techList.contains(IsoDep::class.java.name) -> {
                        standard = "ISO 14443-4 (Type A)"
                        type = "iso7816"
                        val isoDep = IsoDep.get(tag)
                        tagTechnology = isoDep
                        historicalBytes = isoDep.historicalBytes.toHexString()
                    }
                    tag.techList.contains(MifareClassic::class.java.name) -> {
                        standard = "ISO 14443-3 (Type A)"
                        type = "mifare_classic"
                    }
                    tag.techList.contains(MifareUltralight::class.java.name) -> {
                        standard = "ISO 14443-3 (Type A)"
                        type = "mifare_ultralight"
                    }
                    else -> {
                        standard = "ISO 14443-3 (Type A)"
                        type = "unknown"
                    }
                }
            } else if (tag.techList.contains(NfcB::class.java.name)) {
                val bTag = NfcB.get(tag)
                protocolInfo = bTag.protocolInfo.toHexString()
                applicationData = bTag.applicationData.toHexString()
                if (tag.techList.contains(IsoDep::class.java.name)) {
                    type = "iso7816"
                    standard = "ISO 14443-4 (Type B)"
                    val isoDep = IsoDep.get(tag)
                    tagTechnology = isoDep
                    hiLayerResponse = isoDep.hiLayerResponse.toHexString()
                } else {
                    type = "unknown"
                    standard = "ISO 14443-3 (Type B)"
                    tagTechnology = bTag
                }
            } else if (tag.techList.contains(NfcF::class.java.name)) {
                standard = "ISO 18092 (FeliCa)"
                type = "iso18092"
                val fTag = NfcF.get(tag)
                manufacturer = fTag.manufacturer.toHexString()
                systemCode = fTag.systemCode.toHexString()
                tagTechnology = fTag
            } else if (tag.techList.contains(NfcV::class.java.name)) {
                standard = "ISO 15693"
                type = "iso15693"
                val vTag = NfcV.get(tag)
                dsfId = vTag.dsfId.toHexString()
                tagTechnology = vTag
            } else {
                type = "unknown"
                standard = "unknown"
            }

            // detect ndef
            if (tag.techList.contains(Ndef::class.java.name)) {
                val ndefTag = Ndef.get(tag)
                ndefTechnology = ndefTag
                ndefAvailable = true
                ndefType = ndefTag.type
                ndefWritable = ndefTag.isWritable
                ndefCanMakeReadOnly = ndefTag.canMakeReadOnly()
                ndefCapacity = ndefTag.maxSize
            }

            result.success(JSONObject(mapOf(
                    "type" to type,
                    "id" to id,
                    "standard" to standard,
                    "atqa" to atqa,
                    "sak" to sak,
                    "historicalBytes" to historicalBytes,
                    "protocolInfo" to protocolInfo,
                    "applicationData" to applicationData,
                    "hiLayerResponse" to hiLayerResponse,
                    "manufacturer" to manufacturer,
                    "systemCode" to systemCode,
                    "dsfId" to dsfId,
                    "ndefAvailable" to ndefAvailable,
                    "ndefType" to ndefType,
                    "ndefWritable" to ndefWritable,
                    "ndefCanMakeReadOnly" to ndefCanMakeReadOnly,
                    "ndefCapacity" to ndefCapacity
            )).toString())

        }, technologies, options)
    }

    private class MethodResultWrapper internal constructor(result: Result) : Result {

        private val methodResult: Result = result
        private var hasError: Boolean = false;

        companion object {
            // a Handler is always thread-safe, so use a singleton here
            private val handler: Handler by lazy {
                Handler(Looper.getMainLooper())
            }
        }

        override fun success(result: Any?) {
            handler.post {
                ignoreIllegalState {
                    methodResult.success(result)
                }
            }
        }

        override fun error(errorCode: String?, errorMessage: String?, errorDetails: Any?) {
            handler.post {
                ignoreIllegalState {
                    methodResult.error(errorCode, errorMessage, errorDetails)
                }
            }
        }

        override fun notImplemented() {
            handler.post {
                ignoreIllegalState {
                    methodResult.notImplemented()
                }
            }
        }

        private fun ignoreIllegalState(fn: () -> Unit) {
            try {
                if (!hasError) fn()
            } catch (e: IllegalStateException) {
                hasError = true;
                Log.w(TAG, "Exception occurred when using MethodChannel.Result: $e")
                Log.w(TAG, "Will ignore all following usage of object: $methodResult")
            }
        }
    }
}
